src/wp-includes/block-template-utils.php

```diff
@@ -1163,14 +1163,36 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 	}
 
 	if ( ! isset( $query['wp_id'] ) ) {
+		$template_files_query = $query;
 		/*
 		 * If the query has found some user templates, those have priority
 		 * over the theme-provided ones, so we skip querying and building them.
 		 */
-		$query['slug__not_in'] = wp_list_pluck( $query_result, 'slug' );
-		$template_files        = _get_block_templates_files( $template_type, $query );
+		$template_files_query['slug__not_in'] = wp_list_pluck( $query_result, 'slug' );
+		/*
+		 * We need to unset the post_type query param because some templates
+		 * would be excluded otherwise, like `page.html` when looking for
+		 * `page` templates. We need all templates so we can exclude duplicates
+		 * from plugin-registered templates.
+		 * See: https://github.com/WordPress/gutenberg/issues/65584
+		 */
+		unset( $template_files_query['post_type'] );
+		$template_files = _get_block_templates_files( $template_type, $template_files_query );
 		foreach ( $template_files as $template_file ) {
-			$query_result[] = _build_block_template_result_from_file( $template_file, $template_type );
+			// The custom templates with no associated post types are available for all post types.
+			if ( isset( $query['post_type'] ) && ! isset( $template_file['postTypes'] ) ) {
+				$candidate              = _build_block_template_result_from_file( $template_file, $template_type );
+				$default_template_types = get_default_block_template_types();
+				if ( ! isset( $default_template_types[ $candidate->slug ] ) ) {
+					$query_result[] = $candidate;
+				}
+			} elseif (
+				// If the query doesn't specify a post type, or it does and the template has the post type, add it.
+				! isset( $query['post_type'] ) ||
+				( isset( $query['post_type'] ) && isset( $template_file['postTypes'] ) && in_array( $query['post_type'], $template_file['postTypes'], true ) )
```

Comment:
A tiny, "nitpicking" suggestion would be to try to wrap this line if possible for readability reasons. Think this was the proper way to do it:

```suggestion
				(
					isset( $query['post_type'] ) &&
					isset( $template_file['postTypes'] ) &&
					in_array( $query['post_type'], $template_file['postTypes'], true )
				)
```

Reply:
Makes sense! I just committed your suggestion. :+1: